### PR TITLE
Add procedures, change the order of pushes

### PR DESCRIPTION
**Pull request by DesWurstes**
These changes simplify the code

@dagurval Could you too please review the wording? Is there an ambiguity I can clear up in this?

### CODE REVIEW

**Comment by dagurval** [comment]
The changes describing the scripts and push payloads look good to me. I'm not familiar enough with the protocol to comment on the procedure updates.

File: bitcoinfiles.md [hunk 1]
```diff
@@ -81,22 +81,39 @@ The scripts below are designed with the following transaction limits in mind:
...
+- `push 3` is of length `480` bytes
+
+The `vout = 0` output of every upload transaction shall be an `OP_RETURN` push.
+
+To upload, for every transaction:
+- The P2SH outputs are redeemed in the order that increasing `vin` corresponds to increasing data index. (Not applicable to the initial transaction)
+- The `OP_RETURN` output is formed
+- P2SH outputs are formed
+
+The last transaction, redeeming P2SH outputs but not creating newer ones, shall have a metadata-only `OP_RETURN` push.
```
Comment (jcramer) [inline bitcoinfiles.md]: Is there a reason why the final transaction's metadata shouldn't be allowed to contain a data chunk?

File: bitcoinfiles.md [hunk 2]
```diff
@@ -81,22 +81,39 @@ The scripts below are designed with the following transaction limits in mind:
...
 
-The transactions may have OP_RETURN outputs as well. The next transaction's `scriptSig`s must be read before the current transaction's OP_RETURN push. Inputs must be decoded with increasing `vin` indices.
+Padding shall be added to satisfy the conditions that the redeem script above is valid and that every P2SH push is at least `220` bytes. While reading, the bytes after `file_byte_count_int` in the metadata will be discarded.
 
+To read transactions from the latest to the earliest the following procedure is suggested, where `chunks` is a push-to-end sequence of byte arrays:
+- Read the metadata `OP_RETURN` push
+- Append the concatenated P2SH input pushes with increasing index as one element to `chunks`
+- Loop reading previous transactions until the number of transactions given in the metadata is read
```
Comment (jcramer) [inline bitcoinfiles.md:115]: Does the metadata specify number of transactions, or number of chunks?  If I'm interpreting this correctly, with p2sh based files, you can have a maximum of to 4 chunks per transaction?  There would be 1 optional op_return chunk and up to 3 chunks in vout = 1?

File: bitcoinfiles.md [hunk 3]
```diff
@@ -81,22 +81,39 @@ The scripts below are designed with the following transaction limits in mind:
...
 For maximum efficiency, it is suggested that
 
-- `push 1` is of length `480` bytes
-- `push 2` is of length `520` bytes
-- `push 3` is of length `493` bytes (`502` if `signature` is Schnorr-signed)
+- `push 1` is of length `520` bytes
+- `push 2` is of length `493` bytes (`502` if `signature` is Schnorr-signed)
+- `push 3` is of length `480` bytes
+
+The `vout = 0` output of every upload transaction shall be an `OP_RETURN` push.
```
Comment (jcramer) [inline bitcoinfiles.md]: Suggest removing ambiguity by adding something along these lines: 

"P2sh outputs shall follow immediately after the `OP_RETURN` at `vout = 1`, they shall be contiguous, and spent in the same order starting at `vin=0`.  For example, `vout=1` would be spent in `vin=0`, `vout=2` spent to `vin=1`, etc."

A pretty picture might also help users conceptualize, but not required.

**Review by jcramer — COMMENTED** [review]
"Padding shall be added to satisfy the conditions that the redeem script above is valid and that every P2SH push is at least 220 bytes. While reading, the bytes after file_byte_count_int in the metadata will be discarded."

Won't adding padding possibly cause the uploaded file to be corrupted.  Why not just say if there is < X bytes to be pushed then OP_RETURNs must be used?

**Comment by DesWurstes** [comment]
> > "Padding shall be added to satisfy the conditions that the redeem script above is valid and that every P2SH push is at least 220 bytes. While reading, the bytes after file_byte_count_int in the metadata will be discarded."
>
> Won't adding padding possibly cause the uploaded file to be corrupted?

It's always the last bytes to be padded which will be discarded. Padding is a must because:

The reading code can be generalized to reading every byte array that is longer than a signature. So if every push is larger than it, if people use alternative pushing scripts the downloader would still work.

However if there are smaller pushes, other than the downloader problem above, then the standard redeem will be invalid because of empty pushes. There might even be standardness problems due to possible single-byte pushes if we try to distribute the remaining data over remaining pushes.

> then OP_RETURNs must be used?

Because of the problem above. If a P2SH push is being filled that means the transaction already has an OP_RETURN, while the next transaction would have a metadata output. If we try to use OP_RETURN we would need extra code to check 1) Fits in OP_RETURN 2) Fits in the space left in metadata push while the probability below is not significant (for the 0x01 protocol it is 99% that there would be a partial OP_RETURN).

>  Why not just say if there is < X bytes to be pushed then OP_RETURNs must be used?

If we need to push 1000 bytes more that doesn't fit into OP_RETURN; instead of specially handling the case padding has worked well so far over Blockupload.io.

> Is there a reason why the final transaction's metadata shouldn't be allowed to contain a data chunk?

There is a probability of 14% that the remaining data would fit in OP_RETURN (while the rest requiring a P2SH output) assuming the metadata is empty, even less otherwise.

Do you find it worth handling this case? This currently minimizes the number of edge cases.

File: bitcoinfiles.md (hunk 2 above)
Comment (DesWurstes) [reply bitcoinfiles.md:115]: I have calculated that transactions may have up to 50 P2SH inputs, but haven't yet tested that.

I have added "In this protocol, `chunk_count_int` refers to the number of transactions that have `OP_RETURN` data pushes.". Is it OK now?

File: bitcoinfiles.md (hunk 3 above)
Comment (DesWurstes) [reply bitcoinfiles.md]: Updated.

**Comment by jcramer** [comment]
How does a downloader know what is padding and what is bytes associated with the end of the original file.  What if the end of the file is `00..00`?  How will the downloader verify the hash?  

Also, its probably obvious, but just reiterating that the hash256 in the metadata is for the file, not the file with padding added.

**Comment by DesWurstes** [comment]
> How does a downloader know what is padding and what is bytes associated with the end of the original file. What if the end of the file is 00..00? How will the downloader verify the hash?

A buffer is created with length. The order of reading chunks is well-defined. Chunks are copied there. Bytes that don't fit (padding bytes) are dropped. The hash is calculated from this buffer. So the padding bytes don't affect the hash or the final buffer.

> Also, its probably obvious, but just reiterating that the hash256 in the metadata is for the file, not the file with padding added.

Yes!

**EDIT:** Also, I have decided to cancel "Is there a reason why the final transaction's metadata shouldn't be allowed to contain a data chunk?" because the download code allows that already without any changes and but it should be optional to the uploader. Mine now does that, looking forward to opening a Pull Request once https://github.com/simpleledger/BitcoinFilesJS/pull/20 is merged.

**Review by jcramer — APPROVED** [review]

**Pull request merged**